### PR TITLE
React to CheckForOverflowUnderflow in regex source generator

### PR DESCRIPTION
**Pull request by stephentoub**
The regex source generator uses code patterns that might have arithmetic overflows, e.g. a bounds check with `(uint)index < span.Length`.  These are intentional, and they're benign... unless the project/compilation has opted-in to overflow/underflow checking (CheckForOverflowUnderflow).  In that case, the code for many patterns can start throwing false positive overflow exceptions, making the source generator unusable.

This commit causes the generator to look at the CheckOverflow setting in the compilation options, and if it's set, to emit `unchecked { ... }` around all the relevant code.

We'll want to backport this to release/7.0.

https://github.com/dotnet/runtime/issues/78214

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs [hunk 1]
```diff
@@ -36,13 +36,24 @@ public partial class RegexGenerator : IIncrementalGenerator
             "#pragma warning disable CS0219 // Variable assigned but never used",
         };
 
+        private sealed record CompilationData(bool AllowUnsafe = false, bool CheckOverflow = false);
+
         public void Initialize(IncrementalGeneratorInitializationContext context)
         {
+            // To avoid invalidating every regex's output when anything from the compilation changes,
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:43]: @CyrusNajmabadi, can you review this file and make sure I didn't do anything that invalidates your perf efforts?

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs [hunk 2]
```diff
@@ -36,13 +36,24 @@ public partial class RegexGenerator : IIncrementalGenerator
             "#pragma warning disable CS0219 // Variable assigned but never used",
         };
 
+        private sealed record CompilationData(bool AllowUnsafe = false, bool CheckOverflow = false);
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:39]: consider a record-struct.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs [hunk 3]
```diff
@@ -36,13 +36,24 @@ public partial class RegexGenerator : IIncrementalGenerator
...
         public void Initialize(IncrementalGeneratorInitializationContext context)
         {
+            // To avoid invalidating every regex's output when anything from the compilation changes,
+            // we extract from it the only things we care about.
+            IncrementalValueProvider<CompilationData> compilationDataProvider =
+                context.CompilationProvider
+                .Select((x, _) =>
+                    x.Options is CSharpCompilationOptions options ?
+                        new CompilationData(options.AllowUnsafe, options.CheckOverflow) :
+                        new CompilationData());
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:50]: this is fine.

**Review by CyrusNajmabadi — APPROVED** [review]
i dont' see any issue on the IG side.

**Comment by stephentoub** [comment]
> i dont' see any issue on the IG side.

Great, thanks for looking

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 4]
```diff
@@ -172,8 +172,27 @@ static void AppendHashtableContents(IndentedTextWriter writer, IEnumerable<Dicti
         }
 
         /// <summary>Emits the code for the RunnerFactory.  This is the actual logic for the regular expression.</summary>
-        private static void EmitRegexDerivedTypeRunnerFactory(IndentedTextWriter writer, RegexMethod rm, Dictionary<string, string[]> requiredHelpers)
+        private static void EmitRegexDerivedTypeRunnerFactory(IndentedTextWriter writer, RegexMethod rm, Dictionary<string, string[]> requiredHelpers, bool checkOverflow)
         {
+            void EnterCheckOverflow()
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:177]: NIT: no need to reset CI for this, but I think it is a bit more readable in general when we have local funcs at the end of the method (even better if they are after a return statement)

**Comment by joperezr** [comment]
I'm not very familiar with `unchecked` when related to performance so this question might not make sense, but would it make sense to just always emit the `unchecked` block when doing these types of bounds checks even when CheckOverflow is false? Or is the reason we don't want to do that mainly to improve code readability?

**Comment by stephentoub** [comment]
> Or is the reason we don't want to do that mainly to improve code readability?

yup

**Review by joperezr — APPROVED** [review]

**Pull request merged**